From eea951491d62bf0be6b2c3ae95b2fd0d0c36a5bc Mon Sep 17 00:00:00 2001
From: victor-tucci <victor.tucci@beldex.io>
Date: Mon, 12 Jan 2026 05:58:34 +0000
Subject: [PATCH 05/19] UR functions

This commit adds UR functions for UR tasks,
I believe that the right place to get
UR strings is the wallet code itself,
especially because it allows us to
skip the part when we have to store
things to file to encode them later.
Now we are fully in memory
---
 .gitmodules                             |   4 +
 CMakeLists.txt                          |   5 +
 external/CMakeLists.txt                 |   1 +
 external/bc-ur                          |   1 +
 src/wallet/CMakeLists.txt               |   1 +
 src/wallet/api/pending_transaction.cpp  |  32 +++
 src/wallet/api/pending_transaction.h    |   1 +
 src/wallet/api/unsigned_transaction.cpp |  40 ++++
 src/wallet/api/unsigned_transaction.h   |   1 +
 src/wallet/api/wallet.cpp               | 305 ++++++++++++++++++++++++
 src/wallet/api/wallet.h                 |   8 +
 src/wallet/api/wallet2_api.h            |  12 +-
 src/wallet/wallet2.cpp                  |  86 +++++--
 src/wallet/wallet2.h                    |   3 +
 14 files changed, 479 insertions(+), 21 deletions(-)
 create mode 160000 external/bc-ur

diff --git a/.gitmodules b/.gitmodules
index 088e5dba2..d641e29da 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -40,3 +40,7 @@
 [submodule "external/Catch2"]
 	path = external/Catch2
 	url = https://github.com/catchorg/Catch2
+[submodule "external/bc-ur"]
+	path = external/bc-ur
+	url = https://github.com/MrCyjaneK/bc-ur
+	branch = misc
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 617d5ee34..e8dcdf00e 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -68,12 +68,16 @@ if(POLICY CMP0079)
 endif()
 
 set(CMAKE_CXX_STANDARD 17)
+add_definitions(-D_LIBCPP_ENABLE_CXX17_REMOVED_FEATURES) # boost: no template named 'unary_function' in namespace 'std'; did you mean '__unary_function'?
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 set(CMAKE_C_STANDARD 11)
 set(CMAKE_C_STANDARD_REQUIRED ON)
 set(CMAKE_C_EXTENSIONS OFF)
 
+# Add this line to suppress implicit-fallthrough warnings
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-implicit-fallthrough")
+
 set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
 if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.14)
   cmake_policy(SET CMP0083 NEW)
@@ -284,6 +288,7 @@ if(NOT MANUAL_SUBMODULES)
     endfunction ()
     
     message(STATUS "Checking submodules")    
+#    check_submodule(external/bc-ur) 
     check_submodule(external/rapidjson)
     check_submodule(external/trezor-common)
     check_submodule(external/randomx)
diff --git a/external/CMakeLists.txt b/external/CMakeLists.txt
index 83c90e843..c5f3f46ce 100755
--- a/external/CMakeLists.txt
+++ b/external/CMakeLists.txt
@@ -61,6 +61,7 @@ endif()
 
 add_subdirectory(db_drivers)
 add_subdirectory(easylogging++ easyloggingpp)
+add_subdirectory(bc-ur)
 add_subdirectory(randomx EXCLUDE_FROM_ALL)
 add_subdirectory(fmt)
 add_subdirectory(date EXCLUDE_FROM_ALL)
diff --git a/external/bc-ur b/external/bc-ur
new file mode 160000
index 000000000..d82e7c753
--- /dev/null
+++ b/external/bc-ur
@@ -0,0 +1 @@
+Subproject commit d82e7c753e710b8000706dc3383b498438795208
diff --git a/src/wallet/CMakeLists.txt b/src/wallet/CMakeLists.txt
index 1460dc3fa..448700528 100755
--- a/src/wallet/CMakeLists.txt
+++ b/src/wallet/CMakeLists.txt
@@ -40,6 +40,7 @@ add_library(wallet
 target_link_libraries(wallet
   PUBLIC
     multisig
+    bc-ur
     rpc_commands
     rpc_common
     cryptonote_core
diff --git a/src/wallet/api/pending_transaction.cpp b/src/wallet/api/pending_transaction.cpp
index 0ab68c7e0..8b0b55322 100755
--- a/src/wallet/api/pending_transaction.cpp
+++ b/src/wallet/api/pending_transaction.cpp
@@ -42,6 +42,8 @@
 #include <sstream>
 #include <boost/format.hpp>
 
+#include "bc-ur/src/bc-ur.hpp"
+
 namespace Wallet {
 
 EXPORT
@@ -153,6 +155,36 @@ bool PendingTransactionImpl::commit(std::string_view filename_, bool overwrite,
     return good();
 }
 
+EXPORT
+std::string PendingTransactionImpl::commitUR(int max_fragment_length) {
+
+    LOG_PRINT_L3("m_pending_tx size: " << m_pending_tx.size());
+
+    try {
+        std::string ptx = m_wallet.wallet()->dump_tx_to_str(m_pending_tx);
+        good();
+        auto urMessage = ur::string_to_bytes(ptx);
+        ur::ByteVector cbor;
+        ur::CborLite::encodeBytes(cbor, urMessage);
+        std::string type;
+        if (m_wallet.watchOnly()) {
+            type = "xmr-txunsigned";
+        } else {
+            type = "xmr-txsigned";
+        }
+        ur::UR urData = ur::UR(type, cbor);
+        auto encoder = ur::UREncoder(urData, max_fragment_length);
+        std::string output;
+        for(size_t i = 0; i < encoder.seq_len(); i++) {
+            output.append("\n"+encoder.next_part());
+        }
+        return output;
+    } catch (const std::exception &e) {
+        m_status = {Status_Error, std::string(tr("Unknown exception: ")) + e.what()};
+        return "";
+    }
+}
+
 EXPORT
 uint64_t PendingTransactionImpl::amount() const
 {
diff --git a/src/wallet/api/pending_transaction.h b/src/wallet/api/pending_transaction.h
index 6fef89490..5c8b87221 100755
--- a/src/wallet/api/pending_transaction.h
+++ b/src/wallet/api/pending_transaction.h
@@ -48,6 +48,7 @@ public:
     void setError(std::string error_msg) override;
     bool good() const override { return m_status.first == Status_Ok; }
     bool commit(std::string_view filename = "", bool overwrite = false, bool flash = false) override;
+    std::string commitUR(int max_fragment_length = 130) override;
     uint64_t amount() const override;
     uint64_t dust() const override;
     uint64_t fee() const override;
diff --git a/src/wallet/api/unsigned_transaction.cpp b/src/wallet/api/unsigned_transaction.cpp
index 0ddbd7ba6..47c048b06 100755
--- a/src/wallet/api/unsigned_transaction.cpp
+++ b/src/wallet/api/unsigned_transaction.cpp
@@ -40,6 +40,8 @@
 #include <sstream>
 #include <boost/format.hpp>
 
+#include "bc-ur/src/bc-ur.hpp"
+
 namespace Wallet {
 
 EXPORT
@@ -85,6 +87,44 @@ bool UnsignedTransactionImpl::sign(std::string_view signedFileName_)
   return true;
 }
 
+EXPORT
+std::string UnsignedTransactionImpl::signUR(int max_fragment_length)
+{
+    if(m_wallet.watchOnly())
+    {
+        m_status = {Status_Error, tr("This is a watch only wallet")};
+        return "";
+    }
+    std::vector<tools::wallet2::pending_tx> ptx;
+    try
+    {
+        tools::wallet2::signed_tx_set signed_txes;
+        std::string signedTx = m_wallet.wallet()->sign_tx_dump_to_str(m_unsigned_tx_set, ptx, signed_txes);
+        if (signedTx.empty())
+        {
+            m_status = {Status_Error, tr("Failed to sign transaction")};
+            return "";
+        }
+        auto urMessage = ur::string_to_bytes(signedTx);
+        ur::ByteVector cbor;
+        ur::CborLite::encodeBytes(cbor, urMessage);
+        std::string type = "xmr-txsigned";
+        ur::UR urData = ur::UR(type, cbor);
+        auto encoder = ur::UREncoder(urData, max_fragment_length);
+        std::string output;
+        for(size_t i = 0; i < encoder.seq_len(); i++) {
+            output.append("\n"+encoder.next_part());
+        }
+        return output;
+    }
+    catch (const std::exception &e)
+    {
+        m_status = {Status_Error, std::string(tr("Failed to sign transaction")) + e.what()};
+        return "";
+    }
+    return "";
+}
+
 //----------------------------------------------------------------------------------------------------
 EXPORT
 bool UnsignedTransactionImpl::checkLoadedTx(const std::function<size_t()> get_num_txes, const std::function<const wallet::tx_construction_data&(size_t)> &get_tx, const std::string &extra_message)
diff --git a/src/wallet/api/unsigned_transaction.h b/src/wallet/api/unsigned_transaction.h
index 634535abd..06cee707b 100755
--- a/src/wallet/api/unsigned_transaction.h
+++ b/src/wallet/api/unsigned_transaction.h
@@ -53,6 +53,7 @@ public:
     uint64_t txCount() const override;
     // sign txs and save to file
     bool sign(std::string_view signedFileName) override;
+    std::string signUR(int max_fragment_length = 130) override;
     std::string confirmationMessage() const override {return m_confirmationMessage;}
     uint64_t minMixinCount() const override;
 
diff --git a/src/wallet/api/wallet.cpp b/src/wallet/api/wallet.cpp
index e5082e899..3f6d1f576 100755
--- a/src/wallet/api/wallet.cpp
+++ b/src/wallet/api/wallet.cpp
@@ -53,6 +53,7 @@
 #include <unordered_map>
 #include <thread>
 #include "epee/scope_leaver.h"
+#include "bc-ur/src/bc-ur.hpp"
 
 using namespace cryptonote;
 
@@ -1118,6 +1119,25 @@ uint64_t WalletImpl::unlockedBalance(uint32_t accountIndex) const
     return wallet()->unlocked_balance(accountIndex, false);
 }
 
+EXPORT
+uint64_t WalletImpl::viewOnlyBalance(uint32_t accountIndex, const std::vector<std::string> &key_images) const
+{
+    clearStatus();
+
+    std::vector<crypto::key_image> kis;
+    for (const auto &key_image : key_images) {
+        crypto::key_image ki;
+        if (!tools::hex_to_type(key_image, ki))
+        {
+            setStatusError(tr("failed to parse key image"));
+            return 0;
+        }
+        kis.push_back(ki);
+    }
+
+    return wallet()->view_only_balance(accountIndex, kis);
+}
+
 EXPORT
 int WalletImpl::countBns()
 {  
@@ -1313,6 +1333,7 @@ int WalletImpl::autoRefreshInterval() const
     return m_refreshIntervalMillis;
 }
 
+EXPORT
 UnsignedTransaction* WalletImpl::loadUnsignedTx(std::string_view unsigned_filename_) {
   auto unsigned_filename = fs::u8path(unsigned_filename_);
   clearStatus();
@@ -1335,6 +1356,58 @@ UnsignedTransaction* WalletImpl::loadUnsignedTx(std::string_view unsigned_filena
   return transaction;
 }
 
+EXPORT
+UnsignedTransaction *WalletImpl::loadUnsignedTxUR(const std::string &input) {
+    clearStatus();
+    UnsignedTransactionImpl * transaction = new UnsignedTransactionImpl(*this);
+    auto decoder = ur::URDecoder();
+
+    std::string delimiter = "\n";
+    std::string inp = input;
+    size_t pos = 0;
+    std::string token;
+    while ((pos = inp.find(delimiter)) != std::string::npos) {
+        token = inp.substr(0, pos);
+        decoder.receive_part(token);
+        inp.erase(0, pos + delimiter.length());
+    }
+    decoder.receive_part(inp);
+
+    if (decoder.is_failure()) {
+        setStatusError(decoder.result_error().what());
+        transaction->m_status = {UnsignedTransaction::Status::Status_Error, status().second};
+        return transaction;
+    }
+
+    if (!decoder.is_complete()) {
+        setStatusError("file ended but ur didn't complete");
+        transaction->m_status = {UnsignedTransaction::Status::Status_Error, status().second};
+        return transaction;
+    }
+
+    std::string data;
+    auto cbor = decoder.result_ur().cbor();
+    auto i = cbor.begin();
+    auto end = cbor.end();
+    ur::CborLite::decodeBytes(i, end, data);
+
+    if (checkBackgroundSync("cannot load tx") || !wallet()->parse_unsigned_tx_from_str(data, transaction->m_unsigned_tx_set)){
+        setStatusError(tr("Failed to load unsigned transactions"));
+        transaction->m_status = {UnsignedTransaction::Status::Status_Error, status().second};
+        return transaction;
+    }
+
+    // Check tx data and construct confirmation message
+    std::string extra_message;
+    if (!transaction->m_unsigned_tx_set.transfers.second.empty())
+        extra_message = (boost::format("%u outputs to import. ") % (unsigned)transaction->m_unsigned_tx_set.transfers.second.size()).str();
+    transaction->checkLoadedTx([&transaction](){return transaction->m_unsigned_tx_set.txes.size();}, [&transaction](size_t n)->const wallet::tx_construction_data&{return transaction->m_unsigned_tx_set.txes[n];}, extra_message);
+    auto [code, msg] = transaction->status();
+    setStatus(code, std::move(msg));
+
+    return transaction;
+}
+
 EXPORT
 bool WalletImpl::submitTransaction(std::string_view filename_) {
   auto fileName = fs::u8path(filename_);
@@ -1357,6 +1430,61 @@ bool WalletImpl::submitTransaction(std::string_view filename_) {
   return true;
 }
 
+EXPORT
+bool WalletImpl::submitTransactionUR(const std::string &input) {
+    clearStatus();
+    auto decoder = ur::URDecoder();
+
+    std::string delimiter = "\n";
+    std::string inp = input;
+    size_t pos = 0;
+    std::string token;
+    while ((pos = inp.find(delimiter)) != std::string::npos) {
+        token = inp.substr(0, pos);
+        decoder.receive_part(token);
+        inp.erase(0, pos + delimiter.length());
+    }
+    decoder.receive_part(inp);
+
+    if (decoder.is_failure()) {
+        setStatusError(decoder.result_error().what());
+        return false;
+    }
+
+    if (!decoder.is_complete()) {
+        setStatusError("file ended but ur didn't complete");
+        return false;
+    }
+
+    std::string data;
+    auto cbor = decoder.result_ur().cbor();
+    auto i = cbor.begin();
+    auto end = cbor.end();
+    ur::CborLite::decodeBytes(i, end, data);
+    if (checkBackgroundSync("cannot submit tx"))
+        return false;
+    std::unique_ptr<PendingTransactionImpl> transaction(new PendingTransactionImpl(*this));
+
+    bool r = wallet()->parse_tx_from_str(data, transaction->m_pending_tx, NULL);
+    if (!r) {
+        setStatus(Status_Ok, tr("Failed to load transaction from file"));
+        return false;
+    }
+
+    if(!transaction->commit()) {
+        setStatusError(transaction->status().second);
+        return false;
+    }
+
+    return true;
+}
+
+EXPORT
+bool WalletImpl::hasUnknownKeyImages() const
+{
+    return wallet()->has_unknown_key_images();
+}
+
 EXPORT
 bool WalletImpl::exportKeyImages(std::string_view filename_)
 {
@@ -1387,6 +1515,40 @@ bool WalletImpl::exportKeyImages(std::string_view filename_)
   return true;
 }
 
+EXPORT
+std::string WalletImpl::exportKeyImagesUR(size_t max_fragment_length, bool all)
+{
+    if (wallet()->watch_only())
+    {
+        setStatusError(tr("Wallet is view only"));
+        return "";
+    }
+    if (checkBackgroundSync("cannot export key images"))
+        return "";
+
+    try
+    {
+        std::string keyImages = wallet()->export_key_images_str(all);
+        auto urMessage = ur::string_to_bytes(keyImages);
+        ur::ByteVector cbor;
+        ur::CborLite::encodeBytes(cbor, urMessage);
+        ur::UR urData = ur::UR("xmr-keyimage", cbor);
+        auto encoder = ur::UREncoder(urData, max_fragment_length);
+        std::string output;
+        for(size_t i = 0; i < encoder.seq_len(); i++) {
+            output.append("\n"+encoder.next_part());
+        }
+        return output;
+    }
+    catch (const std::exception &e)
+    {
+        LOG_ERROR("Error exporting key images: " << e.what());
+        setStatusError(e.what());
+        return "";
+    }
+    return "";
+}
+
 EXPORT
 bool WalletImpl::importKeyImages(std::string_view filename_)
 {
@@ -1414,6 +1576,61 @@ bool WalletImpl::importKeyImages(std::string_view filename_)
   return true;
 }
 
+EXPORT
+bool WalletImpl::importKeyImagesUR(const std::string &input)
+{
+    if (checkBackgroundSync("cannot import key images"))
+        return false;
+    if (!trustedDaemon()) {
+        setStatusError(tr("Key images can only be imported with a trusted daemon"));
+        return false;
+    }
+    try
+    {
+        auto decoder = ur::URDecoder();
+        std::string delimiter = "\n";
+        std::string inp = input;
+        size_t pos = 0;
+        std::string token;
+        while ((pos = inp.find(delimiter)) != std::string::npos) {
+            token = inp.substr(0, pos);
+            decoder.receive_part(token);
+            inp.erase(0, pos + delimiter.length());
+        }
+        decoder.receive_part(inp);
+
+        if (decoder.is_failure()) {
+            setStatusError(decoder.result_error().what());
+            return false;
+        }
+
+        if (!decoder.is_complete()) {
+            setStatusError("file ended but ur didn't complete");
+            return false;
+        }
+
+        std::string data;
+        auto cbor = decoder.result_ur().cbor();
+        auto i = cbor.begin();
+        auto end = cbor.end();
+        ur::CborLite::decodeBytes(i, end, data);
+
+        uint64_t spent = 0, unspent = 0;
+
+        uint64_t height = wallet()->import_key_images_str(data, spent, unspent);
+        LOG_PRINT_L2("Signed key images imported to height " << height << ", "
+                                                             << print_money(spent) << " spent, " << print_money(unspent) << " unspent");
+    }
+    catch (const std::exception &e)
+    {
+        LOG_ERROR("Error exporting key images: " << e.what());
+        setStatusError(std::string(tr("Failed to import key images: ")) + e.what());
+        return false;
+    }
+
+    return true;
+}
+
 EXPORT
 bool WalletImpl::exportOutputs(const std::string &filename, bool all)
 {
@@ -1448,6 +1665,40 @@ bool WalletImpl::exportOutputs(const std::string &filename, bool all)
     return true;
 }
 
+EXPORT
+std::string WalletImpl::exportOutputsUR(size_t max_fragment_length, bool all)
+{
+
+    if (checkBackgroundSync("cannot export outputs"))
+        return "";
+    if (wallet()->key_on_device())
+    {
+        setStatusError(std::string(tr("Not supported on HW wallets.")));
+        return "";
+    }
+
+    try
+    {
+        std::string data = wallet()->export_outputs_to_str(all);
+        auto urMessage = ur::string_to_bytes(data);
+        ur::ByteVector cbor;
+        ur::CborLite::encodeBytes(cbor, urMessage);
+        ur::UR urData = ur::UR("xmr-output", cbor);
+        auto encoder = ur::UREncoder(urData, max_fragment_length);
+        std::string output;
+        for(size_t i = 0; i < encoder.seq_len(); i++) {
+            output.append("\n"+encoder.next_part());
+        }
+        return output;
+    }
+    catch (const std::exception &e)
+    {
+        LOG_ERROR("Error exporting outputs: " << e.what());
+        setStatusError(std::string(tr("Error exporting outputs: ")) + e.what());
+        return "";
+    }
+}
+
 EXPORT
 bool WalletImpl::importOutputs(const std::string &filename)
 {
@@ -1485,6 +1736,60 @@ bool WalletImpl::importOutputs(const std::string &filename)
     return true;
 }
 
+EXPORT
+bool WalletImpl::importOutputsUR(const std::string &input)
+{
+    if (checkBackgroundSync("cannot import outputs"))
+        return false;
+    if (wallet()->key_on_device())
+    {
+        setStatusError(std::string(tr("Not supported on HW wallets.")));
+        return false;
+    }
+
+    try
+    {
+        auto decoder = ur::URDecoder();
+
+        std::string delimiter = "\n";
+        std::string inp = input;
+        size_t pos = 0;
+        std::string token;
+        while ((pos = inp.find(delimiter)) != std::string::npos) {
+            token = inp.substr(0, pos);
+            decoder.receive_part(token);
+            inp.erase(0, pos + delimiter.length());
+        }
+        decoder.receive_part(inp);
+
+        if (decoder.is_failure()) {
+            setStatusError(decoder.result_error().what());
+            return false;
+        }
+
+        if (!decoder.is_complete()) {
+            setStatusError("file ended but ur didn't complete");
+            return false;
+        }
+
+        std::string data;
+        auto cbor = decoder.result_ur().cbor();
+        auto i = cbor.begin();
+        auto end = cbor.end();
+        ur::CborLite::decodeBytes(i, end, data);
+        size_t n_outputs = wallet()->import_outputs_from_str(std::string(data));
+        LOG_PRINT_L2(std::to_string(n_outputs) << " outputs imported");
+    }
+    catch (const std::exception &e)
+    {
+        LOG_ERROR("Failed to import outputs: " << e.what());
+        setStatusError(std::string(tr("Failed to import outputs: ")) + e.what());
+        return false;
+    }
+
+    return true;
+}
+
 EXPORT
 bool WalletImpl::setupBackgroundSync(const Wallet::BackgroundSyncType background_sync_type, const std::string &wallet_password, const std::optional<std::string> &background_cache_password)
 {
diff --git a/src/wallet/api/wallet.h b/src/wallet/api/wallet.h
index 02f7de267..36f0fab2c 100755
--- a/src/wallet/api/wallet.h
+++ b/src/wallet/api/wallet.h
@@ -121,6 +121,7 @@ public:
     bool setProxy(const std::string &address) override;
     uint64_t balance(uint32_t accountIndex = 0) const override;
     uint64_t unlockedBalance(uint32_t accountIndex = 0) const override;
+    uint64_t viewOnlyBalance(uint32_t accountIndex, const std::vector<std::string> &key_images) const override;
     int countBns() override;
     std::vector<stakeInfo>* listCurrentStakes() const override;
     uint64_t blockChainHeight() const override;
@@ -219,11 +220,18 @@ public:
     std::vector<bnsInfo>* MyBns() const override;
     PendingTransaction* createSweepUnmixableTransaction() override;
     bool submitTransaction(std::string_view filename) override;
+    bool submitTransactionUR(const std::string &input) override;
     UnsignedTransaction* loadUnsignedTx(std::string_view unsigned_filename) override;
+    virtual UnsignedTransaction * loadUnsignedTxUR(const std::string &input) override;
+    bool hasUnknownKeyImages() const override;
     bool exportKeyImages(std::string_view filename) override;
+    std::string exportKeyImagesUR(size_t max_fragment_length, bool all = false) override;
     bool importKeyImages(std::string_view filename) override;
+    bool importKeyImagesUR(const std::string &input) override;
     bool exportOutputs(const std::string &filename, bool all = false) override;
+    std::string exportOutputsUR(size_t max_fragment_length, bool all) override;
     bool importOutputs(const std::string &filename) override;
+    bool importOutputsUR(const std::string &filename) override;
 
     bool setupBackgroundSync(const BackgroundSyncType background_sync_type, const std::string &wallet_password, const std::optional<std::string> &background_cache_password = std::optional<std::string>()) override;
     BackgroundSyncType getBackgroundSyncType() const override;
diff --git a/src/wallet/api/wallet2_api.h b/src/wallet/api/wallet2_api.h
index d8dd26365..f2c54d862 100755
--- a/src/wallet/api/wallet2_api.h
+++ b/src/wallet/api/wallet2_api.h
@@ -76,6 +76,7 @@ struct PendingTransaction
     virtual std::pair<int, std::string> status() const = 0;
     // commit transaction or save to file if filename (utf-8) is provided.
     virtual bool commit(std::string_view filename = "", bool overwrite = false, bool flash = false) = 0;
+    virtual std::string commitUR(int max_fragment_length = 130) = 0;
     virtual uint64_t amount() const = 0;
     virtual uint64_t dust() const = 0;
     virtual uint64_t fee() const = 0;
@@ -156,6 +157,7 @@ struct UnsignedTransaction
     * return - true on success
     */
     virtual bool sign(std::string_view signedFileName) = 0;
+    virtual std::string signUR(int max_fragment_length = 130) = 0;
 };
 
 /**
@@ -645,6 +647,7 @@ struct Wallet
             result += unlockedBalance(i);
         return result;
     }
+    virtual uint64_t viewOnlyBalance(uint32_t accountIndex, const std::vector<std::string> &key_images = {}) const = 0;
 
     /**
     * @brief interface for counting number of BNS
@@ -1014,6 +1017,7 @@ struct Wallet
     *                          after object returned
     */
     virtual UnsignedTransaction* loadUnsignedTx(std::string_view unsigned_filename) = 0;
+    virtual UnsignedTransaction * loadUnsignedTxUR(const std::string &input) = 0;
 
    /*!
     * \brief submitTransaction - submits transaction in signed tx file (utf8 filename)
@@ -1021,7 +1025,7 @@ struct Wallet
     * \return                  - true on success
     */
     virtual bool submitTransaction(std::string_view filename) = 0;
-
+    virtual bool submitTransactionUR(const std::string &input) = 0;
 
     /*!
      * \brief disposeTransaction - destroys transaction object
@@ -1037,12 +1041,15 @@ struct Wallet
      */
     virtual uint64_t estimateTransactionFee(uint32_t priority, uint32_t recipients = 1) const = 0;
 
+    virtual bool hasUnknownKeyImages() const = 0;
+
    /*!
     * \brief exportKeyImages - exports key images to file
     * \param filename - utf8 filename to write to
     * \return                  - true on success
     */
     virtual bool exportKeyImages(std::string_view filename) = 0;
+    virtual std::string exportKeyImagesUR(size_t max_fragment_length, bool all = false) = 0;
 
    /*!
     * \brief importKeyImages - imports key images from file
@@ -1050,6 +1057,7 @@ struct Wallet
     * \return                  - true on success
     */
     virtual bool importKeyImages(std::string_view filename) = 0;
+    virtual bool importKeyImagesUR(const std::string &input) = 0;
 
     /*!
      * \brief exportOutputs - exports outputs to file
@@ -1057,6 +1065,7 @@ struct Wallet
      * \return                  - true on success
      */
     virtual bool exportOutputs(const std::string &filename, bool all = false) = 0;
+    virtual std::string exportOutputsUR(size_t max_fragment_length, bool all = false) = 0;
 
     /*!
      * \brief importOutputs - imports outputs from file
@@ -1064,6 +1073,7 @@ struct Wallet
      * \return                  - true on success
      */
     virtual bool importOutputs(const std::string &filename) = 0;
+    virtual bool importOutputsUR(const std::string &filename) = 0;
 
     /*!
      * \brief setupBackgroundSync       - setup background sync mode with just a view key
diff --git a/src/wallet/wallet2.cpp b/src/wallet/wallet2.cpp
index 27b92a1bd..a3568fa2f 100755
--- a/src/wallet/wallet2.cpp
+++ b/src/wallet/wallet2.cpp
@@ -885,6 +885,16 @@ uint32_t get_subaddress_clamped_sum(uint32_t idx, uint32_t extra)
   return idx + extra;
 }
 
+bool is_preferred_input(const std::vector<crypto::key_image>& preferred_input_list, const crypto::key_image& input) {
+  if (!preferred_input_list.empty()) {
+    auto it = std::find(preferred_input_list.begin(), preferred_input_list.end(), input);
+    if (it == preferred_input_list.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+
 void setup_shim(hw::wallet_shim * shim, tools::wallet2 * wallet)
 {
   shim->get_tx_pub_key_from_received_outs = [wallet] (const auto& td) { return wallet->get_tx_pub_key_from_received_outs(td); };
@@ -6569,6 +6579,25 @@ uint64_t wallet2::unlocked_balance(uint32_t index_major, bool strict, uint64_t *
   return amount;
 }
 //----------------------------------------------------------------------------------------------------
+uint64_t wallet2::view_only_balance(uint32_t index_major, const std::vector<crypto::key_image>& selected_inputs)
+{
+  uint64_t amount = 0;
+  for (const auto &td : m_transfers) {
+    if (is_preferred_input(selected_inputs, td.m_key_image) &&
+        !is_spent(td, false) &&
+        !td.m_frozen &&
+        !td.m_key_image_partial &&
+        td.m_key_image_known &&
+        td.is_rct() &&
+        is_transfer_unlocked(td) &&
+        td.m_subaddr_index.major == index_major)
+    {
+      amount += td.m_amount;
+    }
+  }
+  return amount;
+}
+//----------------------------------------------------------------------------------------------------
 std::map<uint32_t, uint64_t> wallet2::balance_per_subaddress(uint32_t index_major, bool strict) const
 {
   std::map<uint32_t, uint64_t> amount_per_subaddr;
@@ -7739,9 +7768,7 @@ bool wallet2::sign_tx(unsigned_tx_set &exported_txs, std::vector<wallet2::pendin
     crypto::key_derivation derivation;
     std::vector<crypto::key_derivation> additional_derivations;
 
-    // compute public keys from out secret keys
-    crypto::public_key tx_pub_key;
-    crypto::secret_key_to_public_key(txs[n].tx_key, tx_pub_key);
+    crypto::public_key tx_pub_key = get_tx_pub_key_from_extra(tx);
     std::vector<crypto::public_key> additional_tx_pub_keys;
     for (const crypto::secret_key &skey: txs[n].additional_tx_keys)
     {
@@ -11695,7 +11722,7 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
   for (size_t i = 0; i < m_transfers.size(); ++i)
   {
     const transfer_details& td = m_transfers[i];
-    if (!is_spent(td, false) && !td.m_frozen && !td.m_key_image_partial && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
+    if (!is_spent(td, false) && !td.m_frozen && !td.m_key_image_partial && td.m_key_image_known && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
     {
       if (td.amount() > m_ignore_outputs_above || td.amount() < m_ignore_outputs_below)
       {
@@ -11745,9 +11772,15 @@ std::vector<wallet2::pending_tx> wallet2::create_transactions_2(std::vector<cryp
 
   LOG_PRINT_L2("Starting with " << num_nondust_outputs << " non-dust outputs and " << num_dust_outputs << " dust outputs");
 
-  if (unused_dust_indices_per_subaddr.empty() && unused_transfers_indices_per_subaddr.empty())
-    return std::vector<wallet2::pending_tx>();
+  // use tobotoht's code path on view-only wallet, otherwise default to upstream
+  bool throwOnNoEnotes = m_account.get_device().device_protocol() == hw::device::protocol::COLD || m_watch_only || m_multisig || m_is_background_wallet;
 
+  if (throwOnNoEnotes) {
+    THROW_WALLET_EXCEPTION_IF(unused_dust_indices_per_subaddr.empty() && unused_transfers_indices_per_subaddr.empty(), error::wallet_internal_error, "No enotes available to spend");
+  } else {
+    if (unused_dust_indices_per_subaddr.empty() && unused_transfers_indices_per_subaddr.empty())
+      return std::vector<wallet2::pending_tx>();
+  }
   // if empty, put dummy entry so that the front can be referenced later in the loop
   if (unused_dust_indices_per_subaddr.empty())
     unused_dust_indices_per_subaddr.push_back({});
@@ -14278,6 +14311,12 @@ bool wallet2::export_key_images_to_file(const fs::path& filename, bool requested
   // [              ((transfer array offset*) (spend public key) (view public key) {(key image) (signature), ...})                              ]
   // *The offset in the wallet's transfers this exported key image file contains.
 
+  std::string data = export_key_images_str(requested_only);
+  return tools::dump_file(filename, data);
+}
+
+std::string wallet2::export_key_images_str(bool requested_only) const
+{
   PERF_TIMER(__FUNCTION__);
   std::pair<size_t, std::vector<std::pair<crypto::key_image, crypto::signature>>> ski = export_key_images(requested_only);
   const cryptonote::account_public_address &keys = get_account().get_keys().m_account_address;
@@ -14297,7 +14336,7 @@ bool wallet2::export_key_images_to_file(const fs::path& filename, bool requested
   PERF_TIMER(export_key_images_encrypt);
   std::string ciphertext{KEY_IMAGE_EXPORT_FILE_MAGIC};
   ciphertext += encrypt_with_view_secret_key(data);
-  return tools::dump_file(filename, ciphertext);
+  return ciphertext;
 }
 
 //----------------------------------------------------------------------------------------------------
@@ -14358,49 +14397,56 @@ std::pair<size_t, std::vector<std::pair<crypto::key_image, crypto::signature>>>
 
 uint64_t wallet2::import_key_images_from_file(const fs::path& filename, uint64_t& spent, uint64_t& unspent)
 {
-  PERF_TIMER(__FUNCTION__);
   std::string data;
   bool r = tools::slurp_file(filename, data);
 
   THROW_WALLET_EXCEPTION_IF(!r, error::wallet_internal_error, std::string(tr("failed to read file ")) + filename.u8string());
 
+  return import_key_images_str(data, spent, unspent);
+}
+
+uint64_t wallet2::import_key_images_str(const std::string &data, uint64_t &spent, uint64_t &unspent)
+{
+  PERF_TIMER(__FUNCTION__);
+  std::string data_local = data;
+
   if (!tools::starts_with(data, KEY_IMAGE_EXPORT_FILE_MAGIC))
   {
-    THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string("Bad key image export file magic in ") + filename.u8string());
+    THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string("Bad key image export file magic in "));
   }
 
   try
   {
     PERF_TIMER(import_key_images_decrypt);
-    data = decrypt_with_view_secret_key(std::string_view{data}.substr(KEY_IMAGE_EXPORT_FILE_MAGIC.size())).view();
+    data_local = decrypt_with_view_secret_key(std::string_view{data}.substr(KEY_IMAGE_EXPORT_FILE_MAGIC.size())).view();
   }
   catch (const std::exception &e)
   {
-    THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string("Failed to decrypt ") + filename.u8string() + ": " + e.what());
+    THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string("Failed to decrypt ") + ": " + e.what());
   }
 
   const size_t headerlen = 4 + 2 * sizeof(crypto::public_key);
-  THROW_WALLET_EXCEPTION_IF(data.size() < headerlen, error::wallet_internal_error, std::string("Bad data size from file ") + filename.u8string());
+  THROW_WALLET_EXCEPTION_IF(data_local.size() < headerlen, error::wallet_internal_error, std::string("Bad data size from file "));
 
-  uint32_t offset = oxenc::load_little_to_host<uint32_t>(data.data());
+  uint32_t offset = oxenc::load_little_to_host<uint32_t>(data_local.data());
   THROW_WALLET_EXCEPTION_IF(offset > m_transfers.size(), error::wallet_internal_error, "Offset larger than known outputs");
 
   // Validate embedded spend/view public keys
   {
     crypto::public_key public_spend_key, public_view_key;
-    std::memcpy(&public_spend_key, &data[sizeof(offset)], sizeof(public_spend_key));
-    std::memcpy(&public_view_key, &data[sizeof(offset) + sizeof(public_spend_key)], sizeof(public_view_key));
+    std::memcpy(&public_spend_key, &data_local[sizeof(offset)], sizeof(public_spend_key));
+    std::memcpy(&public_view_key, &data_local[sizeof(offset) + sizeof(public_spend_key)], sizeof(public_view_key));
 
     const cryptonote::account_public_address &keys = get_account().get_keys().m_account_address;
     if (public_spend_key != keys.m_spend_public_key || public_view_key != keys.m_view_public_key)
     {
-      THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string( "Key images from ") + filename.u8string() + " are for a different account");
+      THROW_WALLET_EXCEPTION(error::wallet_internal_error, std::string( "Key images from ") + " are for a different account");
     }
   }
 
   const size_t record_size        = sizeof(crypto::key_image) + sizeof(crypto::signature);
-  const size_t record_buffer_size = data.size() - headerlen;
-  THROW_WALLET_EXCEPTION_IF(record_buffer_size % record_size, error::wallet_internal_error, std::string("Bad data size from file ") + filename.u8string());
+  const size_t record_buffer_size = data_local.size() - headerlen;
+  THROW_WALLET_EXCEPTION_IF(record_buffer_size % record_size, error::wallet_internal_error, std::string("Bad data size from file "));
 
   const size_t num_records = record_buffer_size / record_size;
   std::vector<std::pair<crypto::key_image, crypto::signature>> ski(num_records);
@@ -14409,8 +14455,8 @@ uint64_t wallet2::import_key_images_from_file(const fs::path& filename, uint64_t
     size_t const key_image_offset = n * record_size;
     size_t const signature_offset = key_image_offset + sizeof(key_image);
     std::pair<crypto::key_image, crypto::signature> &pair = ski[n];
-    std::memcpy(&pair.first,  &data[headerlen + key_image_offset], sizeof(key_image));
-    std::memcpy(&pair.second, &data[headerlen + signature_offset], sizeof(signature));
+    std::memcpy(&pair.first,  &data_local[headerlen + key_image_offset], sizeof(key_image));
+    std::memcpy(&pair.second, &data_local[headerlen + signature_offset], sizeof(signature));
   }
 
   return import_key_images(ski, offset, spent, unspent);
diff --git a/src/wallet/wallet2.h b/src/wallet/wallet2.h
index 618696e1e..620c2c8ce 100755
--- a/src/wallet/wallet2.h
+++ b/src/wallet/wallet2.h
@@ -899,6 +899,7 @@ private:
     // locked & unlocked balance of given or current subaddress account
     uint64_t balance(uint32_t subaddr_index_major, bool strict) const;
     uint64_t unlocked_balance(uint32_t subaddr_index_major, bool strict, uint64_t *blocks_to_unlock = NULL, uint64_t *time_to_unlock = NULL) const;
+    uint64_t view_only_balance(uint32_t index_major, const std::vector<crypto::key_image>& selected_inputs = {});
     // locked & unlocked balance per subaddress of given or current subaddress account
     std::map<uint32_t, uint64_t> balance_per_subaddress(uint32_t subaddr_index_major, bool strict) const;
     std::map<uint32_t, std::pair<uint64_t, std::pair<uint64_t, uint64_t>>> unlocked_balance_per_subaddress(uint32_t subaddr_index_major, bool strict) const;
@@ -1399,9 +1400,11 @@ private:
     std::tuple<size_t, crypto::hash, std::vector<crypto::hash>> export_blockchain() const;
     void import_blockchain(const std::tuple<size_t, crypto::hash, std::vector<crypto::hash>> &bc);
     bool export_key_images_to_file(const fs::path &filename, bool requested_only) const;
+    std::string export_key_images_str(bool requested_only) const;
     std::pair<size_t, std::vector<std::pair<crypto::key_image, crypto::signature>>> export_key_images(bool requested_only) const;
     uint64_t import_key_images(const std::vector<std::pair<crypto::key_image, crypto::signature>> &signed_key_images, size_t offset, uint64_t &spent, uint64_t &unspent, bool check_spent = true);
     uint64_t import_key_images_from_file(const fs::path& filename, uint64_t &spent, uint64_t &unspent);
+    uint64_t import_key_images_str(const std::string &data, uint64_t &spent, uint64_t &unspent);
     bool import_key_images(std::vector<crypto::key_image> key_images, size_t offset=0, std::optional<std::unordered_set<size_t>> selected_transfers=std::nullopt);
     bool import_key_images(signed_tx_set & signed_tx, size_t offset=0, bool only_selected_transfers=false);
     crypto::public_key get_tx_pub_key_from_received_outs(const transfer_details &td) const;
-- 
2.25.1

